### PR TITLE
Fix for Cucumber Data Table not showing in allure report #7078

### PR DESCRIPTION
**Pull request by praveendvd**
## Proposed changes

image

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue) #6895
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers
@christian-bromann  and @BorisOsipov

### CODE REVIEW

**Comment by praveendvd** [comment]
@christian-bromann  and @BorisOsipov  , how to fix the coverage failures could you help me out here ?

File: packages/wdio-allure-reporter/src/index.ts [hunk 1]
```diff
@@ -126,6 +127,11 @@ class AllureReporter extends WDIOReporter {
     }
 
     onTestStart(test: TestStats | HookStats) {
+        let testObj:any =test
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.ts:130]: Don't use `any`. If you want access `TestStats.argument` add in `TestStats` type

File: packages/wdio-allure-reporter/src/index.ts [hunk 2]
```diff
@@ -126,6 +127,11 @@ class AllureReporter extends WDIOReporter {
...
+        let dataTable = testObj.argument ? testObj.argument.rows.map((a:{cells:null}) => a.cells) : undefined
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.ts:131]: don't use let, prefer `const`

File: packages/wdio-allure-reporter/src/index.ts [hunk 3]
```diff
@@ -126,6 +127,11 @@ class AllureReporter extends WDIOReporter {
...
+        dataTable ? stringify(dataTable, function (err, records) {
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.ts:132]: use stringify sync api https://csv.js.org/stringify/api/#sync-api

File: packages/wdio-allure-reporter/src/index.ts (hunk 1 above)
Comment (praveendvd) [reply packages/wdio-allure-reporter/src/index.ts:130]: This was throwin hookStats doesn't contain arguments error when i `run npm run compile` thats why i provided any

File: packages/wdio-allure-reporter/src/index.ts [hunk 4]
```diff
@@ -126,6 +127,11 @@ class AllureReporter extends WDIOReporter {
...
+            records ? AllureReporter.addAttachment('Data Table', records, 'text/csv') : console.log(err)
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.ts:133]: please don't use static methods of reporter instead use https://github.com/webdriverio/webdriverio/pull/7081/files#diff-c5907206259e7237a9d12ae4b8beff6bd37b92840964c685d194e4cc3233cd75R450

File: packages/wdio-allure-reporter/src/index.ts (hunk 1 above)
Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.ts:130]: Seems here we need some refactoring. Btw I don't see reason to hack typescript types in such a way by using `any`

File: packages/wdio-allure-reporter/src/index.ts (hunk 4 above)
Comment (praveendvd) [reply packages/wdio-allure-reporter/src/index.ts:133]: @BorisOsipov  instead use ? it points to this diff itself . Could you cross check the link

File: packages/wdio-allure-reporter/src/index.ts (hunk 1 above)
Comment (praveendvd) [reply packages/wdio-allure-reporter/src/index.ts:130]: @BorisOsipov  should i add argument to hookstat interface ?

**Comment by BorisOsipov** [comment]
@praveendvd 
>how to fix the coverage failures could you help me out here ?

It is easy. You should write unit tests. I would recomend extract this datatable transformation logic to [utils.ts](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-allure-reporter/src/utils.ts) and write UT for it + UT for cases when `argument` present\absent depends on TestStat object.

**Comment by praveendvd** [comment]
> @praveendvd
> 
> > how to fix the coverage failures could you help me out here ?
> 
> It is easy. You should write unit tests. I would recomend extract this datatable transformation logic to [utils.ts](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-allure-reporter/src/utils.ts) and write UT for it + UT for cases when `argument` present\absent depends on TestStat object.

@BorisOsipov  Thank you so much for the detailed review , will add the changes

File: packages/wdio-allure-reporter/src/index.ts (hunk 4 above)
Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.ts:133]: @praveendvd sure https://github.com/webdriverio/webdriverio/blob/736113e28a369e66898a1d372fc295e6e2726736/packages/wdio-allure-reporter/src/index.ts#L444

File: packages/wdio-allure-reporter/src/index.ts (hunk 1 above)
Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.ts:130]: Not sure. I will be easy to add type guards like
```
        if (test instanceof TestStats) {
            const dataTable = test.argument ? test.argument.rows.map((a:{cells:null}) => a.cells) : undefined
            dataTable ? stringify(dataTable, function (err, records) {
                records ? AllureReporter.addAttachment('Data Table', records, 'text/csv') : console.log(err)}) : null
        }
```

and the same for `argument.rows`

Comment (praveendvd) [reply packages/wdio-allure-reporter/src/index.ts:130]: @BorisOsipov  i used :

```
    onTestStart(test: TestStats | HookStats) {
        const testObj = test as TestStats
        const argument = testObj.argument as Argument
        const dataTable = testObj.argument ? argument.rows.map((a: { cells: null }) => a.cells) : undefined
        if (dataTable) this.addAttachment({ name: 'Data Table', content: stringify(dataTable), type: 'text/csv' })
        const testTitle = test.currentTest ? test.currentTest : test.title
```

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.ts:130]: ok

**Comment by praveendvd** [comment]
@BorisOsipov  Added unit tests and the coverage is increased thanks for the help , moved the addattachment step after startStep as the datable was always getting addded to the last step. Now it gets added to the correct one 

image

**Review by BorisOsipov — APPROVED** [review]

**Comment by praveendvd** [comment]
```
Error: ../../../node_modules/ts-node/dist/repl.d.ts(1,23): error TS4090: Conflicting definitions for 'node' found at '/home/runner/work/webdriverio/webdriverio/node_modules/@types/node/index.d.ts' and '/home/runner/work/webdriverio/webdriverio/packages/webdriverio/node_modules/@types/node/index.d.ts'. Consider installing a specific version of this library to resolve the conflict.
  The file is in the program because:
    Type library referenced via 'node' from file '/home/runner/work/webdriverio/webdriverio/packages/webdriverio/build/types.d.ts' with packageId '@types/node/index.d.ts@15.14.0'
    Type library referenced via 'node' from file '/home/runner/work/webdriverio/webdriverio/node_modules/ts-node/dist/repl.d.ts' with packageId '@types/node/index.d.ts@15.12.5'
```

@BorisOsipov  This error doesn't seems to be an issue with thus pull request , is there something i need to do ?

**Comment by BorisOsipov** [comment]
restart GA checks

**Comment by BorisOsipov** [comment]
@praveendvd yeah it seems not related, but don't know what happened(

**Comment by praveendvd** [comment]
@BorisOsipov  again the same failure :(

**Comment by BorisOsipov** [comment]
@christian-bromann there are some issues in types definitions. We need help to resolve this.

**Comment by christian-bromann** [comment]
@BorisOsipov type issues are resolved

@praveendvd mind rebasing on top of current `main` branch?

**Comment by praveendvd** [comment]
Thank you @christian-bromann and @BorisOsipov for looking into this  I will update the branch and push again

**Comment by praveendvd** [comment]
> @BorisOsipov type issues are resolved
> 
> @praveendvd mind rebasing on top of current `main` branch?

image

@christian-bromann  Thank you Christian for looking into this, all test are passing except for windows:

image

**Comment by christian-bromann** [comment]
mhm .. the current main branch passes.

**Comment by praveendvd** [comment]
@christian-bromann  it passed there was no changes than the package-lock.json , should iremove or keep that file ? if its not required shouldn't be it be part of gitignore ?

**Comment by christian-bromann** [comment]
can you remove it and reinstall dependencies again? I am a bit concerned that we have 20k removed lines and only 36 added ones. Make sure you run Node.js v15 when you install your deps.

**Comment by praveendvd** [comment]
Removed it's passing !! Party

File: packages/wdio-allure-reporter/src/index.ts [hunk 5]
```diff
@@ -134,7 +135,12 @@ class AllureReporter extends WDIOReporter {
         }
 
         if (this._options.useCucumberStepReporter) {
-            return this._allure.startStep(testTitle)
+            const step = this._allure.startStep(testTitle)
+            const testObj = test as TestStats
+            const argument = testObj?.argument as Argument
+            const dataTable = argument?.rows?.map((a: { cells: Array<string> }) => a?.cells)
+            if (dataTable) this._allure.addAttachment('Data Table', stringify(dataTable), 'text/csv')
```
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:142]:
```suggestion
            
            if (dataTable) {
                this._allure.addAttachment('Data Table', stringify(dataTable), 'text/csv')
            }

```

**Review by christian-bromann — APPROVED** [review]
Awesome, thanks a ton 👍

File: packages/wdio-allure-reporter/src/index.ts [hunk 6]
```diff
@@ -134,7 +135,16 @@ class AllureReporter extends WDIOReporter {
         }
 
         if (this._options.useCucumberStepReporter) {
-            return this._allure.startStep(testTitle)
+            const step = this._allure.startStep(testTitle)
+            const testObj = test as TestStats
+            const argument = testObj?.argument as Argument
+            const dataTable = argument?.rows?.map((a: { cells: Array<string> }) => a?.cells)
```
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:141]:
```suggestion
            const dataTable = argument?.rows?.map((a: { cells: string[] }) => a?.cells)
```

File: packages/wdio-allure-reporter/src/index.ts [hunk 7]
```diff
@@ -134,7 +135,16 @@ class AllureReporter extends WDIOReporter {
         }
 
         if (this._options.useCucumberStepReporter) {
-            return this._allure.startStep(testTitle)
+            const step = this._allure.startStep(testTitle)
+            const testObj = test as TestStats
+            const argument = testObj?.argument as Argument
+            const dataTable = argument?.rows?.map((a: { cells: string[] }) => a?.cells)
+
```
Comment (christian-bromann) [inline packages/wdio-allure-reporter/src/index.ts:142]:
```suggestion

```

**Pull request merged**